### ISSUE TITLE
Feature request: support multiple model instances on TensorRT LLM triton backend. 

### ISSUE THREAD

**Issue by wengsnow** [triton-inference-server/tensorrtllm_backend#105]
I used Baichuan2 13B model weight only int 8 and launch a triton server on single GPU. Now I have a node has 2 GPUs and want to multiple model instances to a instance for each GPUs to improve concurrent request throughput . 

For the default config: "kind_cpu count=1", the gpu usage as follows:
![image](https://github.com/triton-inference-server/tensorrtllm_backend/assets/27884697/9eceac52-a8eb-4dd5-96e6-cf9bc458ebcd)

Just like fastertransformer,  I set "kind_cpu count=2" (https://github.com/triton-inference-server/fastertransformer_backend/blob/c6936c39be6fc57c161db34cfe5c45d1bef81ee2/README.md?plain=1#L48)  But it can not work, I get such error:
![image](https://github.com/triton-inference-server/tensorrtllm_backend/assets/27884697/7bef7220-bd28-44e1-9b07-6f9d94e166fa)
When this error happened,  the first GPU memory was increased, the second GPU memory was not changed. 

When I follow the triton guide to set "kind_gpu count=1", get the same error.
(https://github.com/triton-inference-server/server/blob/0f1f9246ee8257db6196da65950205be79379401/docs/user_guide/model_configuration.md?plain=1#L600)

I want to know how to use Multiple Model Instances worked for tensorrtllm, which is a important feature for triton.

**Comment by byshiue** [comment]
TensorRT-LLM does not support multiple model instances now. You can create a issue to request it and we will consider it in our roadmap.

**Comment by wengsnow** [comment]
Thanks for your reply. Could you please tell me how to  improve concurrent request throughput, when I have a node with two A100 40G GPUs.  As shown in the figure 1, only the first GPUs be used for inference, I hope both two GPUs can work.

**Comment by byshiue** [comment]
> Thanks for your reply. Could you please tell me how to improve concurrent request throughput, when I have a node with two A100 40G GPUs. As shown in the figure 1, only the first GPUs be used for inference, I hope both two GPUs can work.

If you only hope to run one instance per GPU, you can run 
```Bash
CUDA_VISIBLE_DEVICES=0 <launch server>
CUDA_VISIBLE_DEVICES=1 <launch server>
```

to run two server on two GPUs. Note that you need to adjust the port to prevent using same port.

**Comment by byshiue** [comment]
The answer is no at the moment.

**Comment by konodyuk** [comment]
@byshiue @ncomly-nvidia 
Could you provide an estimated timeline for implementing multi-instance support? This is a crucial concern for those transitioning from the Fastertransformer backend to TensorRT-LLM.

**Comment by sadreamer** [comment]
Same here. I would also kindly ask for an estimation of when multi-instance will be supported. It's indeed very important for some use cases.
NB in the FT backend there was a similar issue with multi-instance, though: https://github.com/triton-inference-server/fastertransformer_backend/issues/63

**Comment by schetlur-nv** [comment]
We are hoping to have multi-instance in the next month or so.

**Comment by zengruizhao** [comment]
@schetlur-nv hope fixed soon

**Comment by njaramish** [comment]
@schetlur-nv I see that the front page [readme ](https://github.com/triton-inference-server/tensorrtllm_backend/blob/f87ad6bf66c7b7862f8e1ef62a2b93d5c0069989/README.md?plain=1#L351) has been updated to include a --multi-model flag. Does this mean that this feature has been implemented, or is it still in progress?

I attempted to use this option to load multiple TensorRT-LLM models in one Triton server. The models loaded correctly, but I';m getting OOM or segfaults when attempting to issue more than a single inference request at a time (i.e. multiple requests to the same model, or one request to each model). 

I confirmed that the engines for these models are able to service multiple requests when deployed inside their own Triton instance. 

Thanks for your work on this feature!

**Comment by dyoshida-continua** [comment]
@schetlur-nv Is multi-model incompatible with running models with tensor parallelism greater than 1?

**Comment by njaramish** [comment]
Noting that when deploying with `--mulit-model`, the /metrics endpoint loses some important tensorrtllm_backend-specific metrics, including: 
- nv_trt_llm_request_metrics
- nv_trt_llm_runtime_memory_metrics
- nv_trt_llm_kv_cache_block_metrics
- nv_trt_llm_inflight_batcher_metrics
- nv_trt_llm_general_metrics

Would it be possible to include reporting for these metrics when Triton is launched with `--multi-model` enabled?

**Comment by avianion** [comment]
@njaramish did you get the multi model deployment working?

**Comment by njaramish** [comment]
@avianion I was able to get it working with two models that each had a batch size of 1 -- increasing the batch size beyond that caused OOM in some cases and segfault in others.

**Comment by byshiue** [comment]
Since the feature is supported, I will close this issue. For other new issues, please help to create another issue to track and discuss to prevent confusing. Thank you for cooperation.

**Comment by njaramish** [comment]
@byshiue thanks, I'll keep this in mind and create separate issues in the future. 

If others run into an issue similar to the one I described, the issue that I described above is fixed under 0.11.0.dev20240521. @avianion everything works as expected using this version.
